lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,292 @@
+			  currentCandidates.insert(std::pair<Expr, Expr>(rel, cand));
+		  }
+	  }
+  }
+
+  /*
+   * Main loop of Houdini algorithm
+   */
+  void Houdini::runHoudini(HornClauseDB &db, HornifyModule &hm)
+  {
+	  guessCandidate(db);
+
+	  HornClauseDB::RuleVector workList;
+	  workList.insert(workList.end(), db.getRules().begin(), db.getRules().end());
+
+	  while(!workList.empty())
+	  {
+		  HornRule r = workList.front();
+		  errs() << "RULE HEAD: " << *(r.head()) << "\n";
+		  errs() << "RULE BODY: " << *(r.body()) << "\n";
+		  while (validateRule(r, db, hm) == SAT)
+		  {
+			  workList = addUsedRulesBackToWorkList(db, workList, r.head());
+			  weakenRuleHeadCand(r);
+		  }
+		  workList.erase(workList.begin());
+	  }
+  }
+
+  /*
+   * Given a rule, check if it's validate
+   */
+  bool Houdini::validateRule(HornRule r, HornClauseDB &db, HornifyModule &hm)
+  {
+	  Expr ruleHead_cand_app = applyActualArgsToCand(r.head());
+	  Expr neg_ruleHead_cand_app = mk<NEG>(ruleHead_cand_app);
+
+	  Expr ruleBody = r.body();
+	  ExprMap body_map;
+	  ExprVector body_pred_apps;
+	  get_all_pred_apps(ruleBody, db, std::back_inserter(body_pred_apps));
+
+	  for(ExprVector::iterator itr = body_pred_apps.begin(); itr != body_pred_apps.end(); ++itr)
+	  {
+		  body_map.insert(std::pair<Expr, Expr>(*itr, applyActualArgsToCand(*itr)));
+	  }
+
+	  Expr body_cand_app = replace(ruleBody, body_map);
+
+	  Expr whole_cand = mk<AND>(neg_ruleHead_cand_app, body_cand_app);
+
+	  errs() << "WHOLE CANDIDATE: " << *whole_cand << "\n";
+
+	  ZSolver<EZ3> solver(hm.getZContext ());
+	  solver.assertExpr(whole_cand);
+	  solver.toSmtLib(errs());
+	  bool isSat = solver.solve();
+	  if(isSat)
+	  {
+		  errs() << "SAT\n";
+	  }
+	  else
+	  {
+	  	  errs() << "UNSAT\n";
+	  }
+	  return isSat;
+  }
+
+  /*
+   * Given a function application, apply its actual arguments to its candidate
+   */
+  Expr Houdini::applyActualArgsToCand(Expr fapp)
+  {
+	  Expr fdecl = bind::fname(fapp);
+	  Expr cand = currentCandidates[fdecl];
+
+	  ExprMap bvar_map;
+	  ExprVector bvars;
+	  get_all_bvars(cand, std::back_inserter(bvars));
+
+	  for(ExprVector::iterator it = bvars.begin(); it != bvars.end(); ++it)
+	  {
+		  unsigned bvar_id = bind::bvarId(*it);
+		  Expr app_arg = bind::domainTy(fapp, bvar_id);// To improve
+		  bvar_map.insert(std::pair<Expr,Expr>(*it, app_arg));
+	  }
+	  Expr cand_app = replace(cand, bvar_map);
+	  return cand_app;
+  }
+
+  /*
+   * Given a rule, weaken its head's candidate
+   */
+  void Houdini::weakenRuleHeadCand(HornRule r)
+  {
+	  Expr ruleHead_app = r.head();
+	  Expr ruleHead_rel = bind::fname(ruleHead_app);
+	  Expr ruleHead_cand = currentCandidates[ruleHead_rel];
+
+	  if(isOpX<TRUE>(ruleHead_cand))
+	  {
+	  		return;
+	  }
+	  if(!isOpX<AND>(ruleHead_cand))
+	  {
+	  		Expr weaken_cand = mk<TRUE>(ruleHead_cand->efac());
+	  		currentCandidates[ruleHead_rel] = weaken_cand;
+	  }
+	  else
+	  {
+	  		ExprVector head_cand_args;
```

Comment:
Hi Arie,
I pass the expression (B1:INT)<=(B0:INT) to the model.eval(expr) function then it returns (!(B1:INT)<=(B0:INT)). I am confused since model should be the assignment to all variables. So I thought eval () is to apply values to expressions and see the result. But it seems I'm wrong. So I hope to ask what does the eval() do ? Thanks.